PSWebServiceLibrary.php

```diff
@@ -512,9 +526,249 @@ public function delete($options)
+    public function __construct($previous = null)
+    {
+        parent::__construct('Bad Request', 400, $previous);
+    }
+}
+
+/**
+ * @package PrestaShopWebservice
+ */
+class PrestaShopWebserviceUnauthorizedException extends \RuntimeException implements PrestaShopWebserviceException {
+    public function __construct($previous = null)
+    {
+        parent::__construct('Unauthorized', 401, $previous);
+    }
+}
+
+/**
+ * @package PrestaShopWebservice
+ */
+class PrestaShopWebserviceForbiddenException extends \RuntimeException implements PrestaShopWebserviceException {
+    public function __construct($previous = null)
+    {
+        parent::__construct('Forbidden', 403, $previous);
+    }
+}
+
+/**
+ * @package PrestaShopWebservice
+ */
+class PrestaShopWebserviceNotFoundException extends \RuntimeException implements PrestaShopWebserviceException {
+    public  function __construct($previous = null)
+    {
+        parent::__construct('Not Found', 404, $previous);
+    }
+}
+
+/**
+ * @package PrestaShopWebservice
+ */
+class PrestaShopWebserviceMethodNotAllowedException extends \RuntimeException implements PrestaShopWebserviceException {
+    public function __construct($previous = null)
+    {
+        parent::__construct('Method Not Allowed', 405, $previous);
+    }
+}
+
+/**
+ * @package PrestaShopWebservice
+ */
+class PrestaShopWebserviceTooManyRequestsException extends \RuntimeException implements PrestaShopWebserviceException {
+    public function __construct($previous = null)
+    {
+        parent::__construct('Too Many Requests', 429, $previous);
+    }
+}
+
+/**
+ * @package PrestaShopWebservice
+ */
+class PrestaShopWebserviceNoContentException extends \RuntimeException implements PrestaShopWebserviceException {
+    public function __construct($previous = null)
+    {
+        parent::__construct('No Content', 204, $previous);
+    }
+}
+
+/**
+ * @package PrestaShopWebservice
+ */
+class PrestaShopWebserviceStatusException extends \RuntimeException implements PrestaShopWebserviceException {
+    /**
+     * @param int $code
+     * @param string $status
+     * @param ?\Throwable $previous
+     * @return self
+     */
+    private static function shouldNotReceiveStatus($code, $status, $previous = null)
+    {
+        return new self(
+            strtr(
+                'This call to PrestaShop Web Services responded with status `%code% %status%`, this client was not designed to process this kind of response. This can come from your web server or reverse proxy configuration.',
+                array(
+                    '%code%' => $code,
+                    '%status%' => $status,
+                )
+            ),
+            $code,
+            $previous
+        );
+    }
+
+    /**
+     * @param ?\Throwable $previous
+     * @return self
+     */
+    public static function shouldNotReceive100ContinueStatus($previous = null)
+    {
+        return self::shouldNotReceiveStatus(100, 'Continue', $previous);
+    }
```

Comment:
Those methods were created to get the status text, as the client does not provide it in the `$response` array. Not doing so would require to provide a very generic exception message in all those return codes.

All those cases may not be produced by Prestashop's API itself, but they could be produced by the underlying infrastructure and networks between the API and the API client.

Reply:
yeah but you don't need one factory method per message, you can handle it this way:

```php
class PrestaShopWebserviceStatusException extends \RuntimeException implements PrestaShopWebserviceException {
	const STATUS_MESSAGES = array(
		100 => 'Continue',
		101 => 'Switching protocols',
		// ...
	);

    /**
     * @param int $code
     * @param ?\Throwable $previous
     * @return self
     */
    public static function shouldNotReceiveStatus($code, $previous = null)
    {
		$status = (isset(self::STATUS_MESSAGES[$code])) ? self::STATUS_MESSAGES[$code] : 'Nonstandard status';
        return new self(
            strtr(
                'This call to PrestaShop Web Services responded with status `%code% %status%`, this client was not designed to process this kind of response. This can come from your web server or reverse proxy configuration.',
                array(
                    '%code%' => $code,
                    '%status%' => $status,
                )
            ),
            $code,
            $previous
        );
    }
```

Reply:
Having one factory per message is a practice I use, and personal preference making the code self-descriptive with a minimal cyclomatic complexity.

If you prefer the option with the array, I'll change it.

Reply:
Array in constants are available since PHP 5.6, we would need to use static property instead of constant.

Reply:
> Having one factory per message is a practice I use

I think that's generally a good practice when you're throwing exceptions in different scenarios. But in this case, you're just transforming a plethora of error codes into instances of the same exception: you're just transforming an error message from the server into an exception, and only the message changes. When handling a new scenario requires you to change two files to add a new function and a new line in a switch, it's usually a code smell.

> Array in constants are available since PHP 5.6

Honestly I don't think anyone who's running PHP < 7 these days will update anything at all on their stack. I wouldn't be against upping the minimum requirements to PHP 7.1.

Reply:
> [...] in this case, you're just transforming a plethora of error codes into instances of the same exception: you're just transforming an error message from the server into an exception, and only the message changes. When handling a new scenario requires you to change two files to add a new function and a new line in a switch, it's usually a code smell.

I'm sorry but I'm not sure what you are asking me to do to avoid code smells.
Do you want me to use an array of error messages with a unique static factory method?
This will not change anything about the switch statement and the number of files to change.

Reply:
I think what Pablo proposed was to use a generic static function able to build the error messages based on the status code This way you don't need a big switch in `PsWebServiceLibrary::assertStatusCode` the switch is only handled in this exception class And if you use an array to store the messages even the switch part can be avoided

Also I understand the limitation about const in PHP 5.6, what Pablo suggested I think is that maybe we could release a new major version of this library based on PHP 7 to use more modern code As he said people still relying on PHP 5 have a very old environment and are not likely to update their libraries regularly so this wouldn't be a problem